Initial product form field store

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This is a very barebones version of the initial field store/registry, that will allow us to register fields that will be available through the ProductForm fields API, this will also allow plugins to render their own fields to make use off.
I mimicked some of the design implementation from [registering new Gutenberg Blocks](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/README.md), that way this stays somewhat consistent.
Again the above is very barebones for now and will likely be expanded on quite a bit as new requirements are needed.

Partially addresses #36074  .

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

1. Run storybook - `pnpm --filter=storybook storybook`
2. Go to the `product-fields` example -> `WooCommerce Admin/experimental/product-fields`
3. Change between the two fields that are registered and make sure they are updated correctly (again pretty simple)
4. Feel free to edit the storybook as well and add some more visually different fields.

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add initial product fields registry store and helper functions

louwie17 committed:
Add changelog

joelclimbsthings commented on packages/js/components/src/product-fields/README.md:
````diff
@@ -0,0 +1,45 @@
+## Example
+
+```js
+// block.js
````
Is this `block.js` comment from gutenberg?

joelclimbsthings commented on packages/js/components/src/product-fields/README.md:
````diff
@@ -0,0 +1,45 @@
+```js
+// block.js
+( function ( element ) {
+	var el = element.createElement;
````
Might also update this to `const`

joelclimbsthings commented on packages/js/components/src/product-fields/stories/index.tsx:
```diff
@@ -0,0 +1,69 @@
+					</option>
+				) ) }
+			</select>
+			{ renderField( selectedField || fields[ 0 ], { name: 'test' } ) }
```
Would it be slightly simpler to just make `fields[ 0 ]` the default value of `selectedField`? Might just be my habit.

joelclimbsthings commented on packages/js/components/src/product-fields/README.md:
```diff
@@ -0,0 +1,45 @@
+
+registerProductField( 'number', {
+	name: 'number',
+	edit: () => {
```
Do we want a description here somewhere of what `edit()` is meant to do?

joelclimbsthings commented on packages/js/components/src/product-fields/api/registration.ts:
```diff
@@ -0,0 +1,47 @@
+	fieldName: string,
+	settings: ProductFieldDefinition
+) {
+	const name = fieldName;
```
Is there a reason why we're creating another `name` variable here?

joelclimbsthings commented on packages/js/components/src/product-fields/api/registration.ts:
```diff
@@ -0,0 +1,47 @@
+		return;
+	}
+
+	const blockType = {
```
Is this reference to `block` from gutenberg? Also, we don't really need the variable here; could just insert the object below where it's used.

joelclimbsthings reviewed (dismissed):
Great work @louwie17 ! This is working well in this limited test. I left a few minor comments. I also had an issue running storybook initially since I was missing the `rememo` dependency; is that something you meant to add?

louwie17 committed:
Change edit to render and address PR feedback

louwie17 committed:
Make use of memoize-one instead of rememo as memoize-one is already installed

louwie17 commented:
@joelclimbsthings thanks for the feedback, I addressed all your suggestions.
I also end up using `memoize-one` over `rememo` as that was already part of the component package.json and they are practically the same thing.
I didn't necessarily want to install a new package if we plan on moving these out in the near future (potentially).

joelclimbsthings commented:
Great, thanks @louwie17 . Looks good! :shipit:
